File: packages/dsa-ts/src/algorithms/graph/bfs.ts
import type { Graph } from '../../structures/Graph.js';

export interface BFSResult {
  visited:  string[];
  path:     string[] | null;   // null si no hay ruta
  distance: number;
}

/**
 * Breadth-First Search — O(V+E).
 * Encuentra el camino más corto en grafos no ponderados.
 */
export function bfs(graph: Graph, start: string, target?: string): BFSResult {
  if (!graph.hasNode(start)) return { visited: [], path: null, distance: -1 };

  const visited  = new Set<string>();
  const queue:   string[]            = [start];
  const parent   = new Map<string, string | null>([[start, null]]);
  const order:   string[]            = [];

  visited.add(start);

  while (queue.length > 0) {
    const node = queue.shift() as string;
    order.push(node);

    if (node === target) {
      return {
        visited:  order,
        path:     reconstructPath(parent, start, target),
        distance: order.length - 1,
      };
    }

    for (const { to } of graph.neighbors(node)) {
      if (!visited.has(to)) {
        visited.add(to);
        parent.set(to, node);
        queue.push(to);
      }
    }
  }

  return { visited: order, path: null, distance: -1 };
}

function reconstructPath(
  parent: Map<string, string | null>,
  _start: string,
  target: string,
): string[] {
  const path: string[] = [];
  let current: string | null | undefined = target;
  while (current !== null && current !== undefined) {
    path.unshift(current);
    current = parent.get(current);
  }
  return path;
}

File: packages/dsa-py/src/algorithms/graph/bfs.py
"""BFS — O(V+E). Camino más corto en grafos no ponderados."""
from collections import deque

Graph = dict[str, list[tuple[str, float]]]


def bfs(
    graph: Graph,
    start: str,
    target: str | None = None,
) -> dict[str, object]:
    if start not in graph:
        return {"visited": [], "path": None, "distance": -1}

    visited: list[str]          = []
    queue                        = deque([start])
    seen:    set[str]            = {start}
    parent:  dict[str, str | None] = {start: None}

    while queue:
        node = queue.popleft()
        visited.append(node)

        if node == target:
            return {
                "visited":  visited,
                "path":     _reconstruct(parent, target),
                "distance": len(visited) - 1,
            }

        for neighbor, _ in graph.get(node, []):
            if neighbor not in seen:
                seen.add(neighbor)
                parent[neighbor] = node
                queue.append(neighbor)

    return {"visited": visited, "path": None, "distance": -1}


def _reconstruct(parent: dict[str, str | None], target: str) -> list[str]:
    path: list[str] = []
    current: str | None = target
    while current is not None:
        path.append(current)
        current = parent.get(current)
    return list(reversed(path))
